### PR TITLE
Add functions to Multi

### LINKED ISSUE #2508: Allow functions to be given to insert/update/delete/insert_or_update in Ecto.Multi
The function should receive the changes so far and return a struct/changeset. They should be equivalent to calling `run/3`.

We also need to change the module docs for Ecto.Multi to make it clear that we only pre-validate functions that explicitly receive a changeset when the changeset is built.

### PR DESCRIPTION
**Pull request by ericmj**
Closes #2508.

### CODE REVIEW

**Comment by wojtekmach** [comment]
I worked on this recently but never finished it up hence the review :) As far as docs I thought about this:

```diff
   transaction. If any changeset has errors, the transaction won't even be started and the error will be immediately returned.
+
+  Note: `insert/4`, `update/4`, `insert_or_update/4`, and `delete/4`
+  variants that accept a function are not performing such checks.
```

(https://github.com/elixir-ecto/ecto/blob/v2.2.10/lib/ecto/multi.ex#L24)

**Comment by michalmuskala** [comment]
I'm not sure about returning ok/error tuples from the function. We already have one mechanism in changesets to mark error states which is the `valid?` flag - I'm afraid introducing another one here might increase complexity needlessly.

What are the use-cases where the ok/error tuple would be required, over only returning a changeset from the functions?

**Comment by ericmj** [comment]
> What are the use-cases where the ok/error tuple would be required, over only returning a changeset from the functions?

I looked through my use cases and I don't have any need for the ok/error tuple so for me it would be perfectly fine to drop it.

File: lib/ecto/multi.ex [hunk 1]
```diff
@@ -520,4 +545,28 @@ defmodule Ecto.Multi do
           "both Ecto.Multi: #{inspect common}"
     end
   end
+
+  defp operation_fun(:insert_or_update, opts, fun) do
+    fn repo, changes ->
+      case fun.(changes) do
+        {:ok, %Changeset{data: %{__meta__: %{state: :loaded}}} = changeset} ->
+          repo.update(changeset, opts)
```
Comment (josevalim) [inline lib/ecto/multi.ex]: You can just call `repo.insert_or_update`

File: lib/ecto/multi.ex [hunk 2]
```diff
@@ -520,4 +545,28 @@ defmodule Ecto.Multi do
           "both Ecto.Multi: #{inspect common}"
     end
   end
+
+  defp operation_fun(:insert_or_update, opts, fun) do
```
Comment (josevalim) [inline lib/ecto/multi.ex]: I think we don't need this clause as there is a `repo.insert_or_update` :)

**Comment by michalmuskala** [comment]
Ok. So to drive this home, I think we need to:

* drop the ok/error tuple wrapping
* introduce a separate operation, distinct from `run`
* add docs with some examples

**Comment by josevalim** [comment]
Sorry, why do we need a separate operation?
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by michalmuskala** [comment]
I thought it would be easier to have it separate from run. It could also make it at least slightly more inspectable in `Multi.to_list`, but I'm not sure it would be that useful.

**Comment by ericmj** [comment]
It seems a bit overkill to document this since we don't even document what `:run` looks like in the introspection.

**Comment by josevalim** [comment]
@ericmj alright. sounds good to me, ship it.

**Comment by josevalim** [comment]
Actually, don't ship it. There are some comments to be addressed. :)

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**